firedrake/mg/mesh.py

```diff
@@ -201,6 +201,36 @@ def ExtrudedMeshHierarchy(base_hierarchy, layers, kernel=None, layer_height=None
                          refinements_per_level=base_hierarchy.refinements_per_level,
                          nested=base_hierarchy.nested)
 
+def ExtrudedMeshHierarchyUniform(base_hierarchy, base_layers, height, kernel=None,
+                          extrusion_type='uniform', gdim=None,
+                          mesh_builder=firedrake.ExtrudedMesh):
+    """Build a hierarchy of extruded meshes by extruding a hierarchy of meshes.
+
+    :arg base_hierarchy: the unextruded base mesh hierarchy to extrude.
+    :arg mesh_builder: function used to turn a :class:`~.Mesh` into an
+       extruded mesh. Used by pyadjoint.
+
+    See :func:`~.ExtrudedMesh` for the meaning of the remaining parameters.
+    """
+    if not isinstance(base_hierarchy, HierarchyBase):
+        raise ValueError("Expecting a HierarchyBase, not a %r" % type(base_hierarchy))
+    if any(m.cell_set._extruded for m in base_hierarchy):
+        raise ValueError("Meshes in base hierarchy must not be extruded")
+
+    meshes = [mesh_builder(m, base_layers * 2**idx, kernel=kernel,
+                           layer_height=height/(base_layers * 2**idx),
+                           extrusion_type=extrusion_type,
+                           gdim=gdim)
+              for idx, m in enumerate(base_hierarchy._meshes)]
```

Comment:
If refinements_per_level of the base hierarchy is not 1, then this does uniform refinement at the right rate, I think.